### PR TITLE
[autoscaler/k8s] Use ray node's HOME in Kubernetes command runner. 

### LINKED ISSUE #12408: [autoscaler] Support for container images with custom HOME directory in k8s operator
### Describe your feature request

We currently hardcode the `~` expansion to `/home/ray` ([code](https://github.com/ray-project/ray/blob/master/python/ray/autoscaler/_private/constants.py#L50)), which breaks paths with a tilde in containers that have a different `HOME` directory.

A few options here:
- Don't manually expand the tilde ourselves and let bash handle it. I assume there's a reason why we're not doing this already -- @ijrsvt or @ericl do you know?
- Allow users to specify RAY_HOME via an environment variable.
- Explicitly mark it in the docs that we don't support container images that are not built from an official Ray base image.

### PR DESCRIPTION
**Pull request by DmitriGekhtman**
## Why are these changes needed?

Removes `ray_constants.RAY_HOME` from the Kubernetes command runner, instead reading the ray node's `HOME` environment variable. This makes it possible to use an image with any HOME directory set.  
## Related issue number
https://github.com/ray-project/ray/issues/12371
https://github.com/ray-project/ray/issues/11545
Closes https://github.com/ray-project/ray/issues/12408

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

Manually checked that `ray up` `ray rsync down` `ray rsync up` work with Kubernetes cluster config on macOS. More comprehensive tests will be added when https://github.com/ray-project/ray/issues/12371 is fully addressed.

### CODE REVIEW

File: python/ray/autoscaler/_private/command_runner.py [hunk 1]
```diff
@@ -236,16 +234,25 @@ def run_rsync_down(self, source, target, options=None):
...
             self.process_runner.check_call(self.kubectl + [
                 "cp", "{}/{}:{}".format(self.namespace, self.node_id, source),
                 target
             ])
 
     def remote_shell_command_str(self):
-        return "{} exec -it {} bash".format(" ".join(self.kubectl),
-                                            self.node_id)
+        return "{} exec -it {} -- bash".format(" ".join(self.kubectl),
+                                               self.node_id)
```
Comment (ijrsvt) [inline python/ray/autoscaler/_private/command_runner.py:246]: What is this exactly for?

**Review by ijrsvt — APPROVED** [review]
Looks good:
@Leemoonsoo should we also do this for `staroid`?

File: python/ray/autoscaler/_private/command_runner.py (hunk 1 above)
Comment (edoakes) [reply python/ray/autoscaler/_private/command_runner.py:246]: I think this is best practices to separate args from the command

**Review by edoakes — CHANGES_REQUESTED** [review]
Is there a reason we can't embed `printenv HOME` inside the command instead of synchronously executing the command to get the HOME directory? At a minimum we should cache this so we only need to run it once.

**Comment by ijrsvt** [comment]
> Is there a reason we can't embed `printenv HOME` inside the command instead of synchronously executing the command to get the HOME directory? At a minimum we should cache this so we only need to run it once.

Good point! We can do what we are currently doing in the DockerCommandRunner--instance variable for the lifetime.

File: python/ray/autoscaler/_private/command_runner.py (hunk 1 above)
Comment (Gekho457) [reply python/ray/autoscaler/_private/command_runner.py:246]: Also, currently Kubernetes will give you a deprecation warning if you don'y include the double dash.

Comment (ijrsvt) [reply python/ray/autoscaler/_private/command_runner.py:246]: SGTM!

**Comment by Gekho457** [comment]
> > Is there a reason we can't embed `printenv HOME` inside the command instead of synchronously executing the command to get the HOME directory? At a minimum we should cache this so we only need to run it once.
> 
> Good point! We can do what we are currently doing in the DockerCommandRunner--instance variable for the lifetime.

A new Kubernetes command runner is instantiated with each NodeUpdaterThread, so cacheing the result is not too helpful. I'll try embedding $HOME in the command.

**Comment by Gekho457** [comment]
> > > Is there a reason we can't embed `printenv HOME` inside the command instead of synchronously executing the command to get the HOME directory? At a minimum we should cache this so we only need to run it once.
> > 
> > 
> > Good point! We can do what we are currently doing in the DockerCommandRunner--instance variable for the lifetime.
> 
> A new Kubernetes command runner is instantiated with each NodeUpdaterThread, so cacheing the result is not too helpful. I'll try embedding $HOME in the command.

Actually, trying to embed `$HOME` or `printenv HOME` in the rsync/kubectl cp command seems more trouble than it's worth. 
I'll cache the result in case there's ever a situation down the line in which a single instance of the command runner needs the home path more than once.
(I guess that could currently happen if `rsync` fails and it falls back to `kubectl cp`.)

**Comment by edoakes** [comment]
@Gekho457 we can't do something like modify `~/foo/bar` -> `${printenv HOME}/foo/bar`? Is `kubectl` doing something funny here?

**Comment by Gekho457** [comment]
> @Gekho457 we can't do something like modify `~/foo/bar` -> `${printenv HOME}/foo/bar`? Is `kubectl` doing something funny here?

Let's say for simplicity the rsync script has failed and `KubernetesCommandRunner.run_rsync_up`  is falling back to `kubectl cp`. 
You could try 
kubectl -n <namespace> cp <source> pod:${printenv HOME}<destination>
but `${printenv HOME}` would expand to the local home directory instead of the remote one.

**Comment by Gekho457** [comment]
> > @Gekho457 we can't do something like modify `~/foo/bar` -> `${printenv HOME}/foo/bar`? Is `kubectl` doing something funny here?
> 
> Let's say for simplicity the rsync script has failed and `KubernetesCommandRunner.run_rsync_up` is falling back to `kubectl cp`.
> You could try
> `kubectl -n <namespace> cp <source> pod:${printenv HOME}/<destination>`
> but `${printenv HOME}` would expand to the local home directory instead of the remote one.

I guess substituting "${printenv HOME}" for "~" would probably work in `kubectl-rsync.sh` but not in the fall-back `kubectl cp`.

**Comment by ijrsvt** [comment]
@Gekho457 @edoakes FWIW, I've personally tried avoiding inlining any sort of expansion on the DockerCommandRunner side because of shell-expansion issues. Caching it as an instance variable will not persist between `NodeUpdaterThread` creations, **but** will help performance for `ray up`, where this codepath is hit _many_ times.

**Comment by Gekho457** [comment]
In any case, I think it's probably good enough for now.

**Comment by Leemoonsoo** [comment]
> @Leemoonsoo should we also do this for staroid?

@ijrsvt Thanks for ping me :) While staroid extends KubernetesCommandRunner and only overrides the constructor to set a environment variable, I think this change will work for staroid as well.

**Review by edoakes — APPROVED** [review]

**Comment by edoakes** [comment]
@Gekho457 LGTM, but given that there are no CI tests for the k8s autoscaler can you make sure that you've tested this manually before we merge?

**Comment by Gekho457** [comment]
@edoakes Checked that it works on Mac and Linux.

**Pull request merged**